### PR TITLE
refactor out duplicate code

### PR DESCRIPTION
**Pull request by jarkkojs**

### CODE REVIEW

**Comment by jarkkojs** [comment]
@npmccallum I did these changes to factor out any possibilities of `mmledger` failing as I'm debugging SGX2. There's also added self-consistency checks in addition to minimizing the code paths.

**Comment by jarkkojs** [comment]
In SGX we **must** have `mmledger` to detect overlaps because it will be a long time, if ever, we support them. Such is issue needs to be detected when it happens.

By dividing the problem into nicely encapsulated components, it's easy to make the call to `delete()` opt-in without messing the code. It would also permit make merging optional but for  that at least SGX does not have immediate need.

File: src/lib.rs [hunk 1]
```diff
@@ -391,64 +417,141 @@ impl<const N: usize> Ledger<N> {
...
-        region: Region::new(Address::new(0xa000), Address::new(0xd000)),
-        access: Access::empty(),
+    const EMPTY_LEDGER: Ledger<5> = Ledger {
+        records: [
+            Record::EMPTY,
+            Record::EMPTY,
+            Record::EMPTY,
+            Record::EMPTY,
+            Record::EMPTY,
+        ],
```
Comment (npmccallum) [inline src/lib.rs]:
```suggestion
        records: [Record::EMPTY; 5],
```

**Review by haraldh — DISMISSED** [review]

**Review by npmccallum — CHANGES_REQUESTED** [review]

**Comment by jarkkojs** [comment]
@npmccallum please review

**Review by haraldh — DISMISSED** [review]

**Comment by jarkkojs** [comment]
@npmccallum mmap PR is not anymore dependent on this but please approve this when you can because it is needed for continuation.

**Comment by jarkkojs** [comment]
Added snapshot of the Heap to this PR. It can refined from this point forward.

**Comment by jarkkojs** [comment]
moved heap to a separate branch which i upstream later on

**Pull request merged**